Implement Zones inheriting colors from parent Zones as an option.

mcourteaux opened the pull request:
Option for coloring zones without custom color by inheriting it from a parent that has a custom color. If inherited, the color is made slightly darker to distinguish the "sources" of the colors and the ones inheriting from each other.

Option saved to ini file.

Performance considerations:
 
 - parents only fetched when feature enabled _and_ zone-to-be-drawn does not have a custom color.
 - parent array is `alloca()`'d

Screenshot without:

![image](https://github.com/user-attachments/assets/3c3ee730-0aa0-4994-9545-0e825466c05e)

With:

![image](https://github.com/user-attachments/assets/f270379a-105b-45c1-ad56-9345aa199246)

mcourteaux commented:
Closes #884

mcourteaux commented:
Fixed the capitalization in the screenshot: `Inherit Colors` -> `Inherit colors`.
Perhaps I should even change it to `Inherit custom`?

mcourteaux commented:
Another example of this in my jobsystem. I used this to make every job get a different color: 

![image](https://github.com/user-attachments/assets/35d020c6-945f-427a-b249-7544ef407882)

wolfpld commented:
> Fixed the capitalization in the screenshot: `Inherit Colors` -> `Inherit colors`.
> Perhaps I should even change it to `Inherit custom`?

I would name it "Inherit parent colors".

I would also rename `GetZoneParentsUntilTopLevel` to `GetAllZoneParents`, with an assert that the final depth is the depth passed as an parameter.

One more thought: the zone timeline drawing routines are already passing through all the parents. Maybe it would be simpler to just pass the user zone color down to children right there, so that no parent lookup is needed at all?

mcourteaux commented:
> One more thought: the zone timeline drawing routines are already passing through all the parents. Maybe it would be simpler to just pass the user zone color down to children right there, so that no parent lookup is needed at all?

I could do that, but that would require adding a `uint32_t custom_color` field to `struct TimelineDraw`, which you've been trying really hard to make it small it seems.

> I would name it "Inherit parent colors".
> 
> I would also rename `GetZoneParentsUntilTopLevel` to `GetAllZoneParents`, with an assert that the final depth is the depth passed as an parameter.

Will do.

wolfpld commented:
> I could do that, but that would require adding a `uint32_t custom_color` field to `struct TimelineDraw`, which you've been trying really hard to make it small it seems.

The amount of timeline items to draw turns out to be surprisingly low, so it's not a concern.

mcourteaux commented:
It's more annoying to implement than I expected. The utility function `View::GetZoneColor()` is `private` to `View`. So from within `TimelineItemThread::PreprocessZoneLevel()`, I can't use that function to get the color there and populate the `TimelineDraw::customColor` field.

wolfpld commented:
Just change it to `public`.

mcourteaux commented:
Changed it and reverted old changes (rewrote git history).

This new approach also allowed me to implement correctly the color for the Folded zones:

![image](https://github.com/user-attachments/assets/55e64a6f-e575-4941-8159-4214201fe61c)

wolfpld commented on profiler/src/profiler/TracyView_Options.cpp:
```diff
@@ -225,6 +225,9 @@ void View::DrawOptions()
     if( SmallCheckbox( "Ignore custom", &forceColors ) ) m_vd.forceColors = forceColors;
+    ImGui::SameLine();
+    bool inheritColors = m_vd.inheritParentColors;
+    if( SmallCheckbox( "Inherit parent colors", &inheritColors) ) m_vd.inheritParentColors = inheritColors;
```
Add missing space.

mcourteaux commented:
Done 😄

wolfpld commented:
I forgot that zone colors may come from two different sources, so I took a deep dive to understand the code and b359936c04e496c2c5a5a75d4aa5b0c4110aea4e is the result. Please take a look if everything there works as you'd expect.

mcourteaux commented:
You introduced a bug somehow, I believe. Neighboring zones adopt the color as well, randomly. Will investigate:

![image](https://github.com/user-attachments/assets/69d1e69c-0ad8-41fb-a738-9bcefa1be969)

The zone pointed at does not have any color.
